### ISSUE TITLE
Don't use `disabled` on focusable UI controls e.g. "Reset" buttons

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#15524]
Noticed while reviewing #14919 for the Range Control, see https://github.com/WordPress/gutenberg/pull/14919#issuecomment-490796943, and applies to other buttons as well, for example the Font Size picker "Reset" button.

The Reset button in the font size picker:

<img width="895" alt="Screenshot 2019-05-09 at 09 24 31" src="https://user-images.githubusercontent.com/1682452/57438233-bfef0a00-7243-11e9-9dc9-80a2a1b5537e.png">

When using the keyboard and pressing the Reset button, the button gets a `disabled` HTML attribute. Thus, it's not focusable any longer and a focus loss occurs.

Worth noting modern browsers try to implement a smart behavior by the means of a so called [sequential focus navigation starting point](https://bugs.chromium.org/p/chromium/issues/detail?id=454172). Among other things, this feature is meant to help in this kind of situations: pressing Tab again should make the tab order start (more or less) from the place where focus was.

However, this is not guaranteed to work with assistive technologies. Not to mention some browsers don't have this feature (typically IE 11).

To avoid this kind of focus loss, Gutenberg uses a different pattern in a few places. Instead of using the HTML attribute `disabled`:
- noop the button
- use `aria-disabled` to communicate the button inert state
- style the button as it was disabled

This pattern should be extended to all the cases where a UI control that currently has focus gets disabled. Starting from these "Reset" buttons, I'd like to propose to review all the cases where `disabled` is used across the codebase.

**Comment by draganescu** [comment]
Hi @afercia, I re-tested this and I noticed that while the button still gets a disabled attribute, the tabbing order is respected, as in pressing tab moves the focus to line height, just as when the reset button is not disabled.

**Comment by afercia** [comment]
Hi @draganescu. I think you're confusing the so called [sequential focus navigation starting point](https://bugs.chromium.org/p/chromium/issues/detail?id=454172) implemented in modern browsers with the native tab order. As mentioned above, those are two different things. The sequential focus navigation starting point isn't implemented in all browsers and there's no guarantee it works as expected when assistive technologies are in use. To get an idea, you should test with IE 11.

**Comment by youknowriad** [comment]
I don't see any "reset" button on the font size control anymore. So I'm guessing this issue doesn't apply anymore.

**Comment by afercia** [comment]
The font size picker mentioned on this issue is just an example of the pattern that needs to be fixed. More controls do have a 'Reset' button or other buttons that get a `disabled` attribute _while_ they're focused. Using the `disabled` attribute is the pattern that should be checked everywhere.

Specifically, the font size picker now has a [withReset](https://github.com/WordPress/gutenberg/blob/a538973fb340d21a454aee2dabd6f689b012d9df/packages/block-editor/src/components/global-styles/typography-panel.js#L357) property set to false so that it doesn't show the Reset button.  However, [the component still uses the disabled attribute](https://github.com/WordPress/gutenberg/blob/a538973fb340d21a454aee2dabd6f689b012d9df/packages/components/src/font-size-picker/index.tsx#L271-L281), thus potentially triggering a focus loss.

Other examples after a quick search in the codebase:

Combobox control:
https://github.com/WordPress/gutenberg/blob/a538973fb340d21a454aee2dabd6f689b012d9df/packages/components/src/combobox-control/index.tsx#L343-L349

Range control: 
https://github.com/WordPress/gutenberg/blob/a538973fb340d21a454aee2dabd6f689b012d9df/packages/components/src/range-control/index.tsx#L315-L323

Box control:
https://github.com/WordPress/gutenberg/blob/a538973fb340d21a454aee2dabd6f689b012d9df/packages/components/src/box-control/index.tsx#L167-L175

As said earlier, this issue isn't just about the 'Reset' buttons. it is about _any_ control that dynamically gets a `disabled` attribute _while_ it is focused.

**Comment by afercia** [comment]
Specifically to the Font size picker, there's at least one occurrence where it still shows the Reset button:
- Go to the Site Editor
- Select a Group block 
- In the Settings sidebar, go to the Styles tab 
- Observe the Typography panel shows the preset sizes 
- Click 'Set custom size'
- The Typography panel switches to the UI to set a custom size
- Observe there is a 'Reset' button after the Size slider

<img width="600" alt="Screenshot 2023-06-19 at 11 16 06" src="https://github.com/WordPress/gutenberg/assets/1682452/dd750f33-df3f-4f45-aac6-ed350f406109">

**Comment by youknowriad** [comment]
Good catch, I think adding the `__experimentalIsFocusable` to these `Button` components should probably solve the issue.

**Comment by mirka** [comment]
As a categorical issue, I think this can be closed out.

- wp-components `Button` — Has the `accessibleWhenDisabled` prop, and while it's not default true for back compat reasons, we lint for its usage in the repo.
- wp-ui `Button` — Has the `focusableWhenDisabled` prop, which is true by default.
